crates/containerd-shim-wasm/src/sandbox/containerd/client.rs

```diff
@@ -401,90 +377,116 @@ impl Client {
-            _ => {}
-        }
-
+        let image_info = self.get_info(&image_digest)?;
+        let mut needs_precompile =
+            can_precompile && !image_info.labels.contains_key(&precompile_id);
         let layers = manifest
             .layers()
             .iter()
             .filter(|x| is_wasm_layer(x.media_type(), T::supported_layers_types()))
-            .map(|config| self.read_content(config.digest()))
+            .map(|original_config| {
+                let mut digest_to_load = original_config.digest().clone();
+                if can_precompile {
+                    let info = self.get_info(&digest_to_load)?;
+                    if info.labels.contains_key(&precompile_id) {
+                        // Safe to unwrap here since we already checked for the label's existence
+                        digest_to_load = info.labels.get(&precompile_id).unwrap().clone();
+                        log::info!(
+                            "layer {} has pre-compiled content: {} ",
+                            info.digest,
+                            &digest_to_load
+                        );
+                    }
+                }
+                log::debug!("loading digest: {} ", &digest_to_load);
+                self.read_content(&digest_to_load)
+                    .map(|module| WasmLayer {
+                        config: original_config.clone(),
+                        layer: module,
+                    })
+                    .or_else(|e| {
+                        // handle content being removed from the content store out of band
+                        if digest_to_load != *original_config.digest() {
+                            log::error!("failed to load precompiled layer: {}", e);
+                            log::error!("falling back to original layer and marking for recompile");
+                            needs_precompile = can_precompile; // only mark for recompile if engine is capable
+                            self.read_content(original_config.digest())
+                                .map(|module| WasmLayer {
+                                    config: original_config.clone(),
+                                    layer: module,
+                                })
+                        } else {
+                            Err(e)
+                        }
+                    })
+            })
             .collect::<Result<Vec<_>>>()?;
 
-        if layers.is_empty() {
-            log::info!("no WASM modules found in OCI layers");
-            return Ok((vec![], platform));
+        if needs_precompile {
+            log::info!("precompiling layers for image: {}", image.name);
+            match engine.precompile(&layers) {
+                Some(compiled_layer_result) => {
+                    let compiled_layers = compiled_layer_result?;
+
+                    for (i, precompiled_layer) in compiled_layers.iter().enumerate() {
+                        let original_layer = &layers[i];
```

Comment:
This line assumes that all N layers are returned even if only M were precompiled (say a few layers were not wasm).

Reply:
For now, I want avoid handling the situation where multple layers are compiled to 1.  I think this is going to need some bigger changes and would like the design to evolve it to a different api like `process_layers` or `initialize_layers` with more input for folks.  If I use `Result<Vec<Option<WasmLayer>>>` as you suggested, I can avoid looking up the layer here and just skip if `None`

Reply:
https://github.com/containerd/runwasi/issues/504 for tracking updates